P0009/P0009.md

```diff
@@ -1390,6 +1392,12 @@ public:
   // [mdspan.extents.helpers], exposition only helpers
   constexpr size_t @_fwd-prod-of-extents_@(rank_type) const noexcept; // @_exposition only_@
   constexpr size_t @_rev-prod-of-extents_@(rank_type) const noexcept; // @_exposition only_@
+  template<class OtherSizeType>
+    requires(!is_integral_v<OtherSizeType>)
+    static constexpr size_type @_index-cast_@(const OtherSizeType& i) noexcept { return static_cast<size_type>(i); } // @_exposition only_@
```

Comment:
1. The request was to distinguish between ["integer-like" types](http://eel.is/c++draft/iterator.concept.winc#def:integer-like) and other types.  Integer-like types need not be `integral`.
2. The usual way to do constraints is to start with the most general case, and then constrain.  Otherwise, one may run afoul of subsumption.

I suggest not defining this inline, so that we don't have to worry about concepts and subsumption.  Instead, we could just describe this in words.

```c++
template<class T>
static constexpr auto @_index-cast_@(const T& i) noexcept;
```

*Effects:*

    * If `OtherSizeType` is integer-like (**[iterator.concept.winc]**), then equivalent to `return i;`;

    * otherwise, equivalent to `return static_cast<size_type>(i);`.

Reply:
(Integer-class-but-not-integral would be covered by is_convertible_v<Indices, size_type> anyway, so we can actually say integral here.)